Gracefully handle notifications that have no TO address, but do have a CC address

Version: `25fd67ab`

Stacktrace:

```
19:08:03.970 [eListenerContainer-1] DEBUG [                 LoggingAspect] - Dispatching notification to [], cc [notification-demo-cc@jhu.edu] (Notification type: SUBMISSION_SUBMISSION_SUBMITTED, Event URI: http://fcrepo:8080/fcrepo/rest/submissionEvents/6b/88/de/ad/6b88dead-5668-4bb7-b137-5f37b11e56d7, Resource URI: http://fcrepo:8080/fcrepo/rest/submissions/3c/56/75/d9/3c5675d9-6444-4c0b-b522-18e583c2ae4d)
19:08:03.979 [eListenerContainer-1] DEBUG [             CompositeResolver] - Attempting resolution of template value 'file:/templates/submission-submitted-subject.hbr', for named template 'SUBJECT'
19:08:04.018 [eListenerContainer-1] DEBUG [             CompositeResolver] - Attempting resolution of template value 'file:/templates/submission-submitted-body.hbr', for named template 'BODY'
19:08:04.018 [eListenerContainer-1] DEBUG [             CompositeResolver] - Attempting resolution of template value 'file:/templates/footer.hbr', for named template 'FOOTER'
19:08:05.153 [eListenerContainer-1] DEBUG [                 EmailComposer] - Initial recipients: []
19:08:05.153 [eListenerContainer-1] DEBUG [                 EmailComposer] - Applying whitelist to resolved recipients: []
19:08:05.164 [eListenerContainer-1] DEBUG [                 EmailComposer] - Whitelisted recipients: []
19:08:05.167 [eListenerContainer-1] INFO  [                 EmailComposer] - Whitelisted recipients: []
19:08:05.170 [eListenerContainer-1] WARN  [                 LoggingAspect] - FAILED dispatching notification to [], cc [notification-demo-cc@jhu.edu] (Notification type: SUBMISSION_SUBMISSION_SUBMITTED, Event URI: http://fcrepo:8080/fcrepo/rest/submissionEvents/6b/88/de/ad/6b88dead-5668-4bb7-b137-5f37b11e56d7, Resource URI: http://fcrepo:8080/fcrepo/rest/submissions/3c/56/75/d9/3c5675d9-6444-4c0b-b522-18e583c2ae4d)
org.dataconservancy.pass.notification.dispatch.DispatchException: Notification must not have a null or empty to address!
	at org.dataconservancy.pass.notification.dispatch.impl.email.EmailDispatchImpl.dispatch(EmailDispatchImpl.java:91)
	at org.dataconservancy.pass.notification.dispatch.impl.email.EmailDispatchImpl$$FastClassBySpringCGLIB$$eb24defd.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:746)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:56)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.framework.adapter.AfterReturningAdviceInterceptor.invoke(AfterReturningAdviceInterceptor.java:55)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.aspectj.AspectJAfterThrowingAdvice.invoke(AspectJAfterThrowingAdvice.java:62)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
	at org.dataconservancy.pass.notification.dispatch.impl.email.EmailDispatchImpl$$EnhancerBySpringCGLIB$$cac9da5.dispatch(<generated>)
	at org.dataconservancy.pass.notification.impl.DefaultNotificationService.notify(DefaultNotificationService.java:58)
	at org.dataconservancy.pass.notification.app.config.JmsConfig.processMessage(JmsConfig.java:124)
	at sun.reflect.GeneratedMethodAccessor16.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:181)
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.invoke(InvocableHandlerMethod.java:114)
	at org.springframework.jms.listener.adapter.MessagingMessageListenerAdapter.invokeHandler(MessagingMessageListenerAdapter.java:114)
	at org.springframework.jms.listener.adapter.MessagingMessageListenerAdapter.onMessage(MessagingMessageListenerAdapter.java:77)
	at org.springframework.jms.listener.AbstractMessageListenerContainer.doInvokeListener(AbstractMessageListenerContainer.java:736)
	at org.springframework.jms.listener.AbstractMessageListenerContainer.invokeListener(AbstractMessageListenerContainer.java:696)
	at org.springframework.jms.listener.AbstractMessageListenerContainer.doExecuteListener(AbstractMessageListenerContainer.java:674)
	at org.springframework.jms.listener.AbstractPollingMessageListenerContainer.doReceiveAndExecute(AbstractPollingMessageListenerContainer.java:318)
	at org.springframework.jms.listener.AbstractPollingMessageListenerContainer.receiveAndExecute(AbstractPollingMessageListenerContainer.java:257)
	at org.springframework.jms.listener.DefaultMessageListenerContainer$AsyncMessageListenerInvoker.invokeListener(DefaultMessageListenerContainer.java:1189)
	at org.springframework.jms.listener.DefaultMessageListenerContainer$AsyncMessageListenerInvoker.executeOngoingLoop(DefaultMessageListenerContainer.java:1179)
	at org.springframework.jms.listener.DefaultMessageListenerContainer$AsyncMessageListenerInvoker.run(DefaultMessageListenerContainer.java:1076)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.dataconservancy.pass.notification.dispatch.DispatchException: Notification must not have a null or empty to address!
	at org.dataconservancy.pass.notification.dispatch.impl.email.EmailComposer.compose(EmailComposer.java:95)
	at org.dataconservancy.pass.notification.dispatch.impl.email.EmailDispatchImpl.dispatch(EmailDispatchImpl.java:81)
	... 32 common frames omitted
19:08:05.175 [eListenerContainer-1] ERROR [NotificationServiceErrorHandle] - Encountered an unrecoverable error: Listener method 'public void org.dataconservancy.pass.notification.app.config.JmsConfig.processMessage(java.lang.String,java.lang.String,java.lang.String,org.springframework.messaging.Message<java.lang.String>,javax.jms.Message)' threw exception; nested exception is org.dataconservancy.pass.notification.dispatch.DispatchException: Notification must not have a null or empty to address!
org.springframework.jms.listener.adapter.ListenerExecutionFailedException: Listener method 'public void org.dataconservancy.pass.notification.app.config.JmsConfig.processMessage(java.lang.String,java.lang.String,java.lang.String,org.springframework.messaging.Message<java.lang.String>,javax.jms.Message)' threw exception; nested exception is org.dataconservancy.pass.notification.dispatch.DispatchException: Notification must not have a null or empty to address!
	at org.springframework.jms.listener.adapter.MessagingMessageListenerAdapter.invokeHandler(MessagingMessageListenerAdapter.java:122)
	at org.springframework.jms.listener.adapter.MessagingMessageListenerAdapter.onMessage(MessagingMessageListenerAdapter.java:77)
	at org.springframework.jms.listener.AbstractMessageListenerContainer.doInvokeListener(AbstractMessageListenerContainer.java:736)
	at org.springframework.jms.listener.AbstractMessageListenerContainer.invokeListener(AbstractMessageListenerContainer.java:696)
	at org.springframework.jms.listener.AbstractMessageListenerContainer.doExecuteListener(AbstractMessageListenerContainer.java:674)
	at org.springframework.jms.listener.AbstractPollingMessageListenerContainer.doReceiveAndExecute(AbstractPollingMessageListenerContainer.java:318)
	at org.springframework.jms.listener.AbstractPollingMessageListenerContainer.receiveAndExecute(AbstractPollingMessageListenerContainer.java:257)
	at org.springframework.jms.listener.DefaultMessageListenerContainer$AsyncMessageListenerInvoker.invokeListener(DefaultMessageListenerContainer.java:1189)
	at org.springframework.jms.listener.DefaultMessageListenerContainer$AsyncMessageListenerInvoker.executeOngoingLoop(DefaultMessageListenerContainer.java:1179)
	at org.springframework.jms.listener.DefaultMessageListenerContainer$AsyncMessageListenerInvoker.run(DefaultMessageListenerContainer.java:1076)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.dataconservancy.pass.notification.dispatch.DispatchException: Notification must not have a null or empty to address!
	at org.dataconservancy.pass.notification.dispatch.impl.email.EmailDispatchImpl.dispatch(EmailDispatchImpl.java:91)
	at org.dataconservancy.pass.notification.dispatch.impl.email.EmailDispatchImpl$$FastClassBySpringCGLIB$$eb24defd.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:746)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:56)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.framework.adapter.AfterReturningAdviceInterceptor.invoke(AfterReturningAdviceInterceptor.java:55)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.aspectj.AspectJAfterThrowingAdvice.invoke(AspectJAfterThrowingAdvice.java:62)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
	at org.dataconservancy.pass.notification.dispatch.impl.email.EmailDispatchImpl$$EnhancerBySpringCGLIB$$cac9da5.dispatch(<generated>)
	at org.dataconservancy.pass.notification.impl.DefaultNotificationService.notify(DefaultNotificationService.java:58)
	at org.dataconservancy.pass.notification.app.config.JmsConfig.processMessage(JmsConfig.java:124)
	at sun.reflect.GeneratedMethodAccessor16.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:181)
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.invoke(InvocableHandlerMethod.java:114)
	at org.springframework.jms.listener.adapter.MessagingMessageListenerAdapter.invokeHandler(MessagingMessageListenerAdapter.java:114)
	... 10 common frames omitted
Caused by: org.dataconservancy.pass.notification.dispatch.DispatchException: Notification must not have a null or empty to address!
	at org.dataconservancy.pass.notification.dispatch.impl.email.EmailComposer.compose(EmailComposer.java:95)
	at org.dataconservancy.pass.notification.dispatch.impl.email.EmailDispatchImpl.dispatch(EmailDispatchImpl.java:81)
	... 32 common frames omitted
```

--- a/dispatch-impl/src/main/java/org/dataconservancy/pass/notification/dispatch/impl/email/EmailDispatchImpl.java
+++ b/dispatch-impl/src/main/java/org/dataconservancy/pass/notification/dispatch/impl/email/EmailDispatchImpl.java
@@ -25,7 +25,11 @@ import org.simplejavamail.mailer.Mailer;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import javax.mail.Message;
+import java.util.Arrays;
 import java.util.Map;
+
+import static java.lang.String.join;
 
 /**
  * Dispatches {@link Notification}s as email messages.  Email templates are configured by {@link NotificationTemplate}s
@@ -80,6 +84,12 @@ public class EmailDispatchImpl implements DispatchService {
 
             Email email = composer.compose(notification, parameterizedTemplates);
 
+            email.getRecipients().stream()
+                    .filter(r -> Message.RecipientType.TO == r.getType())
+                    .findAny()
+                    .orElseThrow(() ->
+                            new DispatchException("Cannot dispatch email with an empty To: address for notification tuple [" + notificationTuple(notification) + "]", notification));
+
             // send email
 
             mailer.sendMail(email);
@@ -87,9 +97,16 @@ public class EmailDispatchImpl implements DispatchService {
             LOG.trace("Dispatched email with id '{}'", email.getId());
 
             return email.getId();
+        } catch (DispatchException e) {
+            throw e;
         } catch (Exception e) {
             throw new DispatchException(e.getMessage(), e, notification);
         }
+    }
+
+    private static String notificationTuple(Notification notification) {
+        return join(",", Arrays.asList(notification.getResourceUri().toString(),
+                notification.getEventUri().toString()));
     }
 
     Parameterizer getParameterizer() {

--- a/dispatch-impl/src/test/java/org/dataconservancy/pass/notification/dispatch/impl/email/EmailDispatchImplTest.java
+++ b/dispatch-impl/src/test/java/org/dataconservancy/pass/notification/dispatch/impl/email/EmailDispatchImplTest.java
@@ -18,6 +18,7 @@ package org.dataconservancy.pass.notification.dispatch.impl.email;
 import org.apache.commons.io.IOUtils;
 import org.dataconservancy.pass.client.PassClient;
 import org.dataconservancy.pass.model.User;
+import org.dataconservancy.pass.notification.dispatch.DispatchException;
 import org.dataconservancy.pass.notification.model.Notification;
 import org.dataconservancy.pass.notification.model.Notification.Param;
 import org.dataconservancy.pass.notification.model.config.NotificationConfig;
@@ -36,13 +37,16 @@ import java.net.URI;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.UUID;
 import java.util.function.Function;
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyZeroInteractions;
 import static org.mockito.Mockito.when;
 
 /**
@@ -182,4 +186,33 @@ public class EmailDispatchImplTest {
         assertEquals("A Subject", email.getSubject());
         assertEquals(String.join("\n\n", "A Body", "A Footer"), email.getPlainText());
     }
+
+    /**
+     * A nice DispatchException should be thrown if the To field of the email is missing or empty
+     */
+    @Test
+    public void emptyToAddress() {
+        Notification notification = mock(Notification.class);
+        Parameterizer p = mock(Parameterizer.class);
+        EmailComposer c = mock(EmailComposer.class);
+        Email e = mock(Email.class);
+
+        when(notification.getResourceUri()).thenReturn(URI.create(UUID.randomUUID().toString()));
+        when(notification.getEventUri()).thenReturn(URI.create(UUID.randomUUID().toString()));
+
+        when(p.resolveAndParameterize(any(), any())).thenReturn(Collections.emptyMap());
+        when(c.compose(any(), any())).thenReturn(e);
+        when(e.getRecipients()).thenReturn(Collections.emptyList());
+
+        try {
+            underTest = new EmailDispatchImpl(p, mailer, c);
+            underTest.dispatch(notification);
+            fail("Expected Dispatch Exception");
+        } catch (DispatchException expected) {
+            assertTrue(expected.getMessage().contains("dispatch email with an empty To: address"));
+        }
+
+        verifyZeroInteractions(mailer);
+    }
+
 }

